packages/thirdweb/src/react/web/ui/prebuilt/thirdweb/BuyDirectListingButton/index.tsx

```
@@ -0,0 +1,189 @@
+ *
+ *
+ * @param props
+ * @returns <TransactionButton />
+ *
+ * @example
+ * ```tsx
+ * import { BuyDirectListingButton } from "thirdweb/react";
+ *
+ * <BuyDirectListingButton
+ *   contractAddress="0x..." // contract address of the marketplace v3
+ *   chain={...} // the chain which the marketplace contract is deployed on
+ *   client={...} // thirdweb client
+ *   listingId={100n} // the listingId or the item you want to buy
+ *   quantity={1n} // optional - see the docs to learn more
+ * >
+ *   Buy NFT
+ * </BuyDirectListingButton>
+ * ```
+ */
+export function BuyDirectListingButton(props: BuyDirectListingButtonProps) {
+  const {
+    contractAddress,
+    listingId,
+    children,
+    chain,
+    client,
+    quantity,
+    payModal,
+  } = props;
+  const defaultPayModalMetadata = payModal ? payModal.metadata : undefined;
+  const account = useActiveAccount();
+  const contract = getContract({
+    address: contractAddress,
+    client,
+    chain,
+  });
+
+  const { data: payMetadata } = useReadContract(getPayMetadata, {
+    contract,
+    listingId,
+    queryOptions: {
+      enabled: !defaultPayModalMetadata,
+    },
+  });
+
+  const { mutateAsync } = useSendAndConfirmTransaction();
+
+  async function prepareBuyTransaction() {
+    if (!account) {
+      throw new Error("No account detected");
+    }
+    const [listing, { getApprovalForTransaction }, { buyFromListing }] =
+      await Promise.all([
+        getListing({
+          contract,
+          listingId,
+        }),
+        import(
+          "../../../../../../extensions/erc20/write/getApprovalForTransaction.js"
+        ),
+        import(
+          "../../../../../../extensions/marketplace/direct-listings/write/buyFromListing.js"
+        ),
+      ]);
+    if (!listing) {
+      throw new Error(`Could not retrieve listing with ID: ${listingId}`);
+    }
+
+    // For ERC721 listing, always hard-code to 1n
+    const _quantity =
```

Instead of implicitly overriding the inputted quantity here we should throw an error if they try specifying a quantity > 1 for ERC721. Or adjust the types to not allow it. Or both, ideally both to cover both JS and TS

---

wow that's good thinking ser. I will add "cover both js and ts" to my list!

---

although I don't think we can adjust the type? Because we won't know if it's ERC721 until we fetch the listing. So maybe just throw error?

---

@gregfromstl actually do you think it would be better UX + DX to handle the ERC721 case for user? Because in the code we only know if it's ERC721 once we have fetched the listing, and if users intend to buy, it's always gonna be at least "1" anyway